pydocstyle: add page (#12579)
tldr-pages/tldr: pages/common/pydocstyle.md

# pydocstyle

> > A static analysis tool for checking compliance with Python docstring conventions.

Reviewer: ```suggestion
> Statically check Python scripts for compliance with docstring conventions.
```
We start with a verb in the imperative form always when possible.

Author: Done.

> More information: <https://www.pydocstyle.org/en/latest/>.

> - Analyze a Python script file or all the Python scripts inside a specific directory:

Reviewer: ```suggestion
- Analyze a Python script or all the Python scripts inside a specific directory:
```
IMO we don't need the word "file" here

> - Analyze a Python script file or all the Python scripts inside a specific directory:

Reviewer 2: Agreed

> - Analyze a Python script file or all the Python scripts inside a specific directory:

Author: Done.

`pydocstyle {{file.py|path/to/directory}}`

> - Show [e]xplanation of each error:

Author: Done.

> `pydocstyle -e | --explain {{file.py|path/to/directory}}`

Author: Done.

- Show [d]ebug information:

> `pydocstyle -d | --debug {{file.py|path/to/directory}}`

Author: Done.

> - Count and print the total number of errors to stdout:

Reviewer: ```suggestion
- Display the total number of errors:
```
I think this way leaves the output I/O stream (`stdout` in this case) implicit, resulting in a more concise description.

> - Count and print the total number of errors to stdout:

Author: Hello @vitorhcl, thank you for the review. All the time you MUST be explicit. 

> "Explicit is better than implicit."

Ref: https://en.m.wikipedia.org/wiki/Zen_of_Python#Principles

For this case I will take your suggestion, the user can use this guide as a "proxy" to the "real" documentation.

> - Count and print the total number of errors to stdout:

Author: Done.

`pydocstyle --count {{file.py|path/to/directory}}`

> - Use the given config file and disable config discovery:

Author: Done.

> `pydocstyle --config={{path/to/config_file}} {{file.py|path/to/directory}}`

Author: Done.

> - Specified which errors to ignore out of all of the available error codes (with a list of comma-separated error codes or prefixes). for example: `--ignore=D101,D2`:

Reviewer: ```suggestion
- Ignore one or more errors:
```
We can be even more direct, some time ago I did some PRs that removed "specify" on some pages when it was possible to avoid it and also left "a comma-separated list" implicit, being demonstrated in the list placeholder.

> - Specified which errors to ignore out of all of the available error codes (with a list of comma-separated error codes or prefixes). for example: `--ignore=D101,D2`:

Reviewer 2: I agree

> - Specified which errors to ignore out of all of the available error codes (with a list of comma-separated error codes or prefixes). for example: `--ignore=D101,D2`:

Author: Hello @tricantivu and @vitorhcl.

Same as above, I will opt for the more simplistic description.

> - Specified which errors to ignore out of all of the available error codes (with a list of comma-separated error codes or prefixes). for example: `--ignore=D101,D2`:

Author: Done.

`pydocstyle --ignore {{D101,D2,D107,...}} {{file.py|path/to/directory}}`

- Check for errors from a specific convention:

> `pydocstyle --convention {{pep257|numpy|google}} {{file.py|directory}}`

Reviewer 2: ```suggestion
`pydocstyle --convention {{pep257|numpy|google}} {{file.py|/path/to/directory}}`
```

Reviewer: ```suggestion
`pydocstyle --convention {{pep257|numpy|google}} {{file.py|path/to/directory}}`
```
Is the path absolute? If not, we should remove the starting slash character

Reviewer 2: Can be both AFAIK

Reviewer: Yeah, I also think it can be both. From the style guide:
> - When the path cannot be relative,
>   but has to start at the root of the filesystem,
>   prefix it with a slash,
>   such as `get {{/path/to/remote_file}}`.

Reviewer 2: I still do not understand why  we should use relative pathnames in this scenario, maybe because its a use case that suits the tool? (e.g. scanning a file from the source directory)

Author: Hello @vitorhcl,

It should be as your suggestion (in order to follow the examples).

Author: @tricantivu 

I am using a "relative" path instead of an "aboslute" path becuase my examples trying to explain that `pydocstyle` can be used also at the "project" context.

```
.
├── README.md
├── app
│   ├── producer.py
│   └── requirements.txt
├── setup.cfg
└── tests
    ├── consumer_test.py
    ├── domain_a
    └── domain_b
        ├── class
        │   ├── data.py
        │   └── terminal.py
        └── translator
            └── translator.py
```

```
# Analyze only the `tests/domain_b/` directory.
pydocstyle tests/domain_b/`
```

Is this answer your question?

Reviewer: Yeah, otherwise we would need to specify like $HOME/project_path/tests/domain_b/

Author: At the end, both suggestion can live together, but I preffer to keep using the previous one (not `/` at the start) in order to have the same example syntax.

Reviewer 2: > @tricantivu 
> 
> 
> 
> I am using a "relative" path instead of an "aboslute" path becuase my examples trying to explain that `pydocstyle` can be used also at the "project" context.
> 
> 
> 
> ```
> 
> .
> 
> ├── README.md
> 
> ├── app
> 
> │   ├── producer.py
> 
> │   └── requirements.txt
> 
> ├── setup.cfg
> 
> └── tests
> 
>     ├── consumer_test.py
> 
>     ├── domain_a
> 
>     └── domain_b
> 
>         ├── class
> 
>         │   ├── data.py
> 
>         │   └── terminal.py
> 
>         └── translator
> 
>             └── translator.py
> 
> ```
> 
> 
> 
> ```
> 
> # Analyze only the `tests/domain_b/` directory.
> 
> pydocstyle tests/domain_b/`
> 
> ```
> 
> 
> 
> Is this answer your question?

Yes, it does.

> `pydocstyle --ignore=D101,D2 {{file.py|path/to/directory}}`

Reviewer: ```suggestion
`pydocstyle --ignore={{D101,D2,D107,...}} {{file.py|path/to/directory}}`
```

Author: Done.

> - Choose the basic list of checked errors by specifying an existing convention. Possible conventions: `pep257`, `numpy`, `google`:

Author: Done.

> `pydocstyle --convention={{name}} {{file.py|directory}}`

Reviewer: ```suggestion
`pydocstyle --convention={{pep257|numpy|google}} {{file.py|directory}}`
```

Author: Done, please resolve the thread if you looks good for you.
